### PR TITLE
Fix nested in generic type binding issues in enabled nullability context

### LINKED ISSUE #61171: Using a more verbose qualification of the identifier of the return type of a property results in name conflict-related errors
**Version Used**: Version 17.2.0 Preview 6.0

## Example

**EDIT**: This bug is consistently reproduced in the context of enabled nullability (`#nullable enable` or `<Nullable>enable</Nullable>`). With nullability disabld in that context, this bug does not occur.

Consider the abstract property:
```csharp
public abstract NodeType NodeType { get; }
```
where `NodeType` is declared within `Tree<TValue>`, so it can be referenced as `Tree<TValue>.NodeType`.

When autocompleting the property's override, the return type is written as the most verbose version of the nested type, even if `NodeType` is visible in that scope:
```csharp
public override Tree<TValue>.NodeType NodeType => throw new NotImplementedException();
```

Then, when using the simplified version of the nested type's name that conflicts with the property's name, an error will be yielded, as it will consider the right-hand `NodeType` to be the property name, instead of the returning type's name.
```csharp
public override Tree<TValue>.NodeType NodeType => NodeType.Leaf;
```

**Expected Behavior**:
- Shorten the name of the return type as much as possible when generating the override's code, if visible in that scope.
- Show a suggestion that the name of the inner type can be shortened (in the above example to `NodeType`).
- Allow writing:
```csharp
public override Tree<TValue>.NodeType NodeType => NodeType.Leaf;
```
for the above example.

**Actual Behavior**:
None of the above.

[Update(jcouv):] [sharplab repro](https://sharplab.io/#v2:EYLgtghglgdgPgYhgVwDaosVBTABNmTHAWACgABAZlwGdsIcATXcgJlwBUAnbbAHg4A1BsmwA+MgG8yuWS2qYaAFy4QAxkpbsAcgHtG2GXOmk5Z+bkUr1mvQY4BPAA5472Ry9yTcAc2xKAblwAXyNZUNIwizoGbGY2XAAFCB4YJTdcEFw3KJNzOSpcXQA3bC4uKANs/XdnVxqPPABeMWr7OoA6ZNTAqIiowpimLVwAGXoAMwysnNNjKLNCkrKKqu5eAWFUUTEOt0a22s8Ww8aO8YgJgIW5AHpb3ABRct0uS2BlrPzvn9+zAD9AUDAX0yANqARkGBTnVcjdZN0CEoADTwsaTVFzcJkYJAA===)

### PR DESCRIPTION
**Pull request by AlFasGD**
Closes #61171

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -6046,7 +6046,7 @@ private BoundExpression BindLeftIdentifierOfPotentialColorColorMemberAccess(Iden
                         {
                             var typeDiagnostics = BindingDiagnosticBag.Create(diagnostics);
                             var boundType = BindNamespaceOrType(left, typeDiagnostics);
-                            if (TypeSymbol.Equals(boundType.Type, leftType, TypeCompareKind.ConsiderEverything2))
+                            if (TypeSymbol.Equals(boundType.Type, leftType, TypeCompareKind.AllNullableIgnoreOptions))
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:6049]: Here is an example with object vs. dynamic:
```
public class Tree<TValue>
{
    public enum NodeType
    {
        Parent,
        Leaf,
    }
}

public class Tree1 : Tree<object>
{
}

public class Tree2 : Tree1
{
    public sealed class LeafNode
    {
        public Tree<dynamic>.NodeType NodeType => NodeType.Leaf;
    }
}
```

Similarly, custom modifiers can be thrown into the mixture, by defining ```Tree``` and ```Tree1```in IL, adding modifiers to the ```object``` type argument in the base type of ```Tree1``` and using ```Tree<dynamic>.NodeType``` as property type.

I agree that we should use ```AllIgnoreOptions``` here.

[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:6049]: Perfect, thanks Aleksey. @AlFasGD, please add a test with the above code (it's expected to compile), and give me a ping on discord and I can walk you through adding the custom modifier version of the test.

[the remark this answers is not in the document]
Comment (Neme12) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:6049]: @333fred Should there also be a test with tuples added?
```c#
public class Tree<TValue>
{
    public enum NodeType
    {
        Parent,
        Leaf,
    }
}

public class Tree1 : Tree<(int a, int b)>
{
}

public class Tree2 : Tree1
{
    public sealed class LeafNode
    {
        public Tree<(int, int)>.NodeType NodeType => NodeType.Leaf;
    }
}
```
This fails to compile as well.

[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Yup.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ColorColorTests.cs [hunk 2]
```diff
@@ -2182,5 +2182,159 @@ class C2
...
+    .method public hidebysig specialname rtspecialname 
+        instance void .ctor () cil managed 
+    {
+        // Method begins at RVA 0x2050
+        // Code size 7 (0x7)
+        .maxstack 8
+
+        IL_0000: ldarg.0
+        IL_0001: call instance void [mscorlib]System.Object::.ctor()
+        IL_0006: ret
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ColorColorTests.cs:2311]:
```suggestion
        ldarg.0
        call instance void [mscorlib]System.Object::.ctor()
        ret
```

File: src/Compilers/CSharp/Test/Semantic/Semantics/ColorColorTests.cs [hunk 3]
```diff
@@ -2182,5 +2182,159 @@ class C2
...
+    .method public hidebysig specialname rtspecialname 
+        instance void .ctor () cil managed 
+    {
+        // Method begins at RVA 0x2058
+        // Code size 7 (0x7)
+        .maxstack 8
+
+        IL_0000: ldarg.0
+        IL_0001: call instance void class Tree`1<object>::.ctor()
+        IL_0006: ret
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ColorColorTests.cs:2329]:
```suggestion
        ldarg.0
        call instance void class Tree`1<object>::.ctor()
        ret
```

**Review by 333fred — APPROVED** [review]
LGTM (commit 3), with a small suggestion on reducing the IL size in the tests.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for the second review of this community PR.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

**Comment by jcouv** [comment]
Thanks @AlFasGD for the fix! Will go ahead with merge/squash as soon as CI is green.

**Pull request merged**